airflow/settings.py

```diff
@@ -71,6 +74,42 @@ def timing(cls, stat, dt):
 
 Stats = DummyStatsLogger
 
+
+def validate_stat(f):
+    @wraps(f)
+    def wrapper(stat, *args, **kwargs):
+        try:
+            validate_key(stat)
+        except Exception as err:
+            log.warning('Invalid stat name: {stat}.'.format(stat=stat), err)
+            return
+        return f(stat, *args, **kwargs)
+
+    return wrapper
+
+
+class SafeStatsdLogger(object):
+
+    def __init__(self, statsd_client):
+        self.statsd = statsd_client
+
+    @validate_stat
```

Comment:
I am kinda confused on the valid / invalid meaning here for stats. Based onhttps://github.com/apache/airflow/blob/f8a24aaa94bbb6361510d58ae367d7e119c44e45/airflow/models/__init__.py#L2111, the `validate_key` is to validate the task_id name, but the stat name should be generic which shouldn't limit to task_id in this case.

Are we agreed upon that stat name should follow the definition in `validate_key` as well?

Reply:
Think twice and I think that we should decouple the stats name rule from the task/dag id rule as some stats store have event stricter rule. We should also allow user to define a custom method to handle characters not allowed.